upstairs/src/volume.rs

```diff
@@ -0,0 +1,1228 @@
+
+        // TODO migration task to pull blocks in from parent, then set parent to
+        // None when done - read block by block, don't overwrite if owned by sub
+        // volume
+
+        self.read_only_parent = Some(SubVolume {
+            // Read only parent LBA range always starts from 0
+            lba_range: Range {
+                start: 0,
+                end: number_of_blocks,
+            },
+            block_io,
+        });
+
+        Ok(())
+    }
+
+    pub fn add_read_only_parent_create_guest(
+        &mut self,
+        opts: CrucibleOpts,
+        gen: u64,
+    ) -> Result<(), CrucibleError> {
+        let guest = Arc::new(Guest::new());
+
+        // Spawn crucible tasks
+        let guest_clone = guest.clone();
+        tokio::spawn(async move {
+            // XXX result eaten here!
+            let _ = up_main(opts, guest_clone).await;
+        });
+
+        guest.activate(gen)?;
+
+        self.add_read_only_parent(guest)
+    }
+
+    // Imagine three sub volumes:
+    //
+    //  0 -> Range(0, a)
+    //  1 -> Range(a, b)
+    //  2 -> Range(b, c)
+    //
+    // Read or write requests could come in and require one, two, or three:
+    //
+    //                   0            a            b            c
+    //   sub volumes:    |------------|------------|------------|
+    //   read request 1:     |------|
+    //   read request 2:                   |------------|
+    //   read request 3:    |-------------------------------|
+    //
+    // For request 1, it affects sub volume 0, and the offset that the user
+    // wants maps directly to the offset of the sub volume.
+    //
+    // For request 2, it affects sub volumes 1 and 2. The logical block address
+    // of the volume has to be translated to the logical block address of
+    // the sub volume.
+    //
+    // For request 3, it affects all sub volumes.
+    //
+    // Sort that out here. Note: start and length are in blocks!
+    //
+    pub fn sub_volumes_for_lba_range(
+        &self,
+        start: u64,
+        length: u64,
+    ) -> Vec<(Range<u64>, &SubVolume)> {
+        let mut sv_vec = vec![];
+
+        for sub_volume in &self.sub_volumes {
+            let coverage = sub_volume.lba_range_coverage(start, length);
+            if let Some(coverage) = coverage {
+                sv_vec.push((coverage, sub_volume));
+            }
+        }
+
+        sv_vec
+    }
+
+    pub fn read_only_parent_for_lba_range(
+        &self,
+        start: u64,
+        length: u64,
+    ) -> Option<Range<u64>> {
+        // Does the read only parent apply for this range
+        if let Some(ref read_only_parent) = self.read_only_parent {
+            read_only_parent.lba_range_coverage(start, length)
+        } else {
+            None
+        }
+    }
+}
+
+impl BlockIO for Volume {
+    fn activate(&self, gen: u64) -> Result<(), CrucibleError> {
+        for sub_volume in &self.sub_volumes {
+            sub_volume.conditional_activate(gen)?;
```

Comment:
I added conditional activation, but we may want to scrap this, and instead have the regular activate function not throw an UpstairsAlreadyActive error?

Reply:
We have to handle the case where one upstairs activates and the other does not.  A 2nd attempt to activate will then fail on the first upstairs (because it's already active).  So we will have to navigate these sorts of issues.

Reply:
Issue #197 covers this